### PR TITLE
Add installation documentation

### PR DESCRIPTION
**Pull request by rkevin-arch**
Still haven't finished rewriting the course management script documentation, but I wanted to get it done before 8AM. I'll finish cleaning up the documentation for course management later today.

### CODE REVIEW

File: docs/user_guide/install_jupyterhub.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
+Intalling in a regular JupyterHub environment as a managed service
+==================================================================
+
+This guide assumes you already know how to set up a JupyterHub environment. You should also be familiar with [adding JupyterHub-managed services](https://jupyterhub.readthedocs.io/en/stable/reference/services.html#hub-managed-services) into ``jupyterhub_config.py``.
```
Comment (lxylxy123456) [inline docs/user_guide/install_jupyterhub.rst:4]:
```suggestion
This guide assumes you already know how to set up a JupyterHub environment. You should also be familiar with `adding JupyterHub-managed services <https://jupyterhub.readthedocs.io/en/stable/reference/services.html#hub-managed-services>`_ into ``jupyterhub_config.py``.
```

File: docs/user_guide/cmdline.rst [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+You should not be using these command-line arguments unless you know what you're doing or have a very specific need (such as running ngshare as an external service).
+
+``--debug``
+^^^^^^^^^^^
+Enable debug mode. This gives more helpful error messages and enables features like dumping the database. **WARNING:** Enabling this *will* leak private information, do *NOT* turn this on in production.
+
+``--no-upgrade-db``
+^^^^^^^^^^^^^^^^^^^
+Do not use Alembic to automatically upgrade the ngshare database. This will cause ngshare to break after an update if the database schema has changed.
```
Comment (lxylxy123456) [inline docs/user_guide/cmdline.rst:33]: Add more warning or provide a reference to admin notes?

**Comment by lxylxy123456** [comment]
Also add this if you have time: "uninstalling ngshare"

File: docs/user_guide/install.rst [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
-===============
+Installing ngshare
+==================
 
-This section is under construction
+``ngshare`` is designed to be installed on a Z2JH cluster, but you may install it without Kubernetes.
 
+.. toctree::
+   :maxdepth: 2
+   :caption: Install ngshare
```
Comment (lxylxy123456) [inline docs/user_guide/install.rst:8]:
```suggestion
```
No need to repeat the title

File: docs/user_guide/install_z2jh.rst [hunk 4]
```diff
@@ -0,0 +1,120 @@
+Installing on a Z2JH cluster
+============================
+
+This guide assumes you are already familiar with installing Z2JH. You should also be familiar with using Helm.
+
+If you prefer looking at examples instead, `here's <https://github.com/lxylxy123456/ngshare/testing/install_z2jh>`_ a sample installation setup. It doesn't demonstrate all the configurable options, though.
```
Comment (lxylxy123456) [inline docs/user_guide/install_z2jh.rst:6]: Broken link (likely to be broken even if merged)

File: docs/user_guide/install_jupyterhub.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+If you prefer looking at examples instead, `here's <https://github.com/lxylxy123456/ngshare/testing/install_jhmanaged>`_ a sample installation setup. It doesn't demonstrate all the configurable options, though.
```
Comment (lxylxy123456) [inline docs/user_guide/install_jupyterhub.rst:6]: Broken link (likely to be broken even if merged)

File: docs/user_guide/install_z2jh.rst [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+Finally, you need to configure nbgrader to use ngshare_exchange. This can be done by adding the following to nbgrader's global config file, ``/etc/jupyter/nbgrader_config.py``:
+
+.. code:: python
+
+    from ngshare_exchange import configureExchange
+    c=get_config()
+    configureExchange(c)
+    c.CourseDirectory.course_id = '*'
+
+A sample singleuser Dockerfile that does all of the above is available `on Github <https://github.com/lxylxy123456/ngshare/testing/install_z2jh/Dockerfile-singleuser>`_.
```
Comment (lxylxy123456) [inline docs/user_guide/install_z2jh.rst:118]: Broken link (likely to be broken even if merged)

File: docs/user_guide/notes_instructor.rst [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+Course cretation
+----------------
+Only the administrators can create courses due to security concerns. Please contact your system administrator if you want to create a course. After they assign you as an instructor, you may manage the course roster and add more students to the course yourself.
+
+Managing students
+-----------------
+Please use the `ngshare-course-management <course_management.html>`_ tool when adding / removing students from a course. **Do not use Formgrader's interface to add students**, since this does not update ngshare.
+
+Configuring formgrader
```
Comment (Lawrence37) [inline docs/user_guide/notes_instructor.rst:13]:
```suggestion
Configuring Formgrader
```

File: testing/install_jhmanaged/nbgrader_config.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from ngshare_exchange import configureExchange
+c=get_config()
+# Note: It's important to specify the right ngshare URL when not using k8s
+configureExchange(c, 'http://127.0.0.1:10101/services/ngshare')
+c.CourseDirectory.course_id = '*'
```
Comment (Lawrence37) [inline testing/install_jhmanaged/nbgrader_config.py:5]: Same thing here. Add a comment explaining what '*' does.

File: docs/user_guide/cmdline.rst [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+======================
+
+Here's a list of command line arguments you may specify when starting ``ngshare``.
+
+Regular arguments
+-----------------
+
+``--database PATH_TO_DATABASE``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Specify a custom database for SQLAlchemy. Defaults to ``sqlite:////srv/ngshare/ngshare.db``. Note that using other types of databases (such as mysql) is not tested.
```
Comment (aalmanza1998) [inline docs/user_guide/cmdline.rst:11]:
```suggestion
Specify a custom database for SQLAlchemy. Defaults to ``sqlite:////srv/ngshare/ngshare.db``. Note that using other types of databases (such as MySQL) is not tested.
```

File: docs/user_guide/cmdline.rst [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+``--storage PATH_TO_STORAGE``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Specify a folder to store user-uploaded files. Defaults to ``/srv/ngshare/files/``.
+
+``--admins ADMIN1,ADMIN2,ADMIN3``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Specify usernames of administrators, separated by commas. Administrators may create courses and access any course.
```
Comment (aalmanza1998) [inline docs/user_guide/cmdline.rst:19]:
```suggestion
Specify usernames of administrators separated by commas. Administrators may create courses and access any course.
```

Could be wrong here but I don't think we need that comma.

File: docs/user_guide/extra.rst [hunk 11]
```diff
@@ -44,3 +44,16 @@ Health Endpoint
...
+
+vngshare stands for Vserver-like Notebook Grader Share. It is similar to `vserver <https://github.com/lxylxy123456/ngshare-vserver/>`_ and allows easy testing.
+
+To run vngshare, do the following:
+
+1. Install dependencies. ``pip3 install tornado jupyterhub sqlalchemy``
+2. ``cd ngshare``
+3. Run vngshare. ``python3 vngshare.py [--host <bind_IP_address> [--port <port_number>]]``
+
+vngshare will create a database at ``/tmp/ngshare.db``. Though there is no file system API, unauthorized users can corrupt your data. You can test vngshare by running it with the default IP and port and executing ``pytest test_ngshare.py``.
```
[the remark this answers is not in the document]
Comment (rkevin-arch) [reply docs/user_guide/extra.rst:59]: vngshare.rst has been deleted and the contents are moved over

File: docs/user_guide/notes_instructor.rst [hunk 12]
```diff
@@ -2,24 +2,28 @@ Notes for Instructors
 =====================
 Make sure to read the following to understand how to manage courses with ``ngshare``.
 
-Course cretation
+Course Creation
 ----------------
```
Comment (Lawrence37) [inline docs/user_guide/notes_instructor.rst:6]:
```suggestion
---------------
```

File: docs/user_guide/extra.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (Lawrence37) [reply docs/user_guide/extra.rst:59]: Still here: https://github.com/lxylxy123456/ngshare/blob/installation_documentation/docs/contributer_guide/vngshare.rst

**Pull request merged**